arangod/Agency/Agent.cpp

```diff
@@ -416,10 +433,7 @@ priv_rpc_ret_t Agent::recvAppendEntriesRPC(term_t term, std::string const& leade
       LOG_TOPIC("b0b19", DEBUG, Logger::AGENCY)
           << "Finished empty AppendEntriesRPC from " << leaderId
           << " with term " << term;
-      {
-        WRITE_LOCKER(oLocker, _outputLock);
-        _commitIndex = std::max(_commitIndex, std::min(leaderCommitIndex, lastIndex));
-      }
+      _commitIndex = std::max(_commitIndex.load(std::memory_order_relaxed), std::min(leaderCommitIndex, lastIndex));
```

Comment:
I don't think this is thread-safe now.
We are not holding any locks here (intentionally), but we do read from `_commitIndex` and write into it in a non-atomic fashion.
If multiple threads do this at the same time, one thread could clobber the update of the other.

Reply:
A lock-free solution could look like this:
```
auto commitIndex = _commitIndex.load();
while (true) {
  auto newCommitIndex = std::max(commitIndex, std::min(leaderCommitIndex, lastIndex));
  if (newCommitIndex == commitIndex || _commitIndex.compare_exchange_weak(commitIndex, newCommitIndex)) {
    break;
  }
}
```

Reply:
Oh this is fine. The _commitIndex on the followers recvAppendEntries() is only eventually consistent.

Reply:
Looking at this for 15 minutes now I tend to agree that this is right. However, I know that the implementation in 3.7 and devel is different, namely, we do acquire the `_outputLock` on principle to change `_commitIndex`, since that lock protects the invariant that `_readDB` and `_commitIndex` refer to the same Raft index. This is follower code here, so there is no `_readDB`, so there is nothing to protect. But this is hard to understand and easy to forget and makes a mental proof that everything is OK more difficult. Therefore, for the sake of better understandability I vote for adopting the solution from 3.7 and devel, then everything should be clear:

```
      {
        WRITE_LOCKER(oLocker, _outputLock);
        auto ci = _commitIndex.load(std::memory_order_relaxed);
        _commitIndex =  std::max(ci , std::min(leaderCommitIndex, lastIndex));
      }
```

Reply:
@kvahed @neunhoef Could one of you take another look here?

Reply:
If there are no protests, I will simply port the `devel` version into this branch and run the tests again.

Reply:
Done. Let's wait for Jenkins - again.